internal/s3fs/s3fs.go

```diff
@@ -0,0 +1,257 @@
+package s3fs
+
+import (
+	"context"
+	"errors"
+	"io"
+	"io/fs"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/aws/aws-sdk-go-v2/service/s3/types"
+	"go.uber.org/zap"
+)
+
+type S3ClientAPI interface {
+	ListObjectsV2(context.Context, *s3.ListObjectsV2Input, ...func(*s3.Options)) (*s3.ListObjectsV2Output, error)
+	GetObject(context.Context, *s3.GetObjectInput, ...func(*s3.Options)) (*s3.GetObjectOutput, error)
+}
+
+// FS is only for accessing files in a single bucket. The directory
+// entries are cached. It is specifically intended for use by a source
+// that calls fs.WalkDir and does not fully implement all fs operations
+type FS struct {
+	logger   *zap.Logger
+	s3Client S3ClientAPI
+
+	bucket   string
+	dirEntry *Dir
+	entries  []fs.DirEntry
+}
+
+// ensure FS implements fs.FS aka Open
+var _ fs.FS = &FS{}
+
+// ensure FS implements fs.StatFS aka Stat
+var _ fs.StatFS = &FS{}
+
+// ensure FS implements fs.ReadDirFS aka ReadDir
+var _ fs.ReadDirFS = &FS{}
+
+// New creates a FS for the single bucket
+func New(logger *zap.Logger, s3Client S3ClientAPI, bucket string) (*FS, error) {
+	return &FS{
+		logger:   logger,
+		s3Client: s3Client,
+		bucket:   bucket,
+	}, nil
+}
+
+// Open implements fs.FS. For the S3 filesystem, it fetches the object
+// contents from s3.
+func (f *FS) Open(name string) (fs.File, error) {
+	if name == "." {
+		return f.dirEntry, nil
+	}
+	pathError := &fs.PathError{
+		Op:   "Open",
+		Path: name,
+		Err:  fs.ErrNotExist,
+	}
+	if !fs.ValidPath(name) {
+		pathError.Err = fs.ErrInvalid
+		return nil, pathError
+	}
+
+	output, err := f.s3Client.GetObject(context.Background(),
+		&s3.GetObjectInput{
+			Bucket: &f.bucket,
+			Key:    &name,
+		})
+	if err != nil {
+		// try to return fs compatible error if possible
+		var nsbe *types.NoSuchBucket
+		if errors.As(err, &nsbe) {
+			return nil, pathError
+		}
+		var nske *types.NoSuchKey
+		if errors.As(err, &nske) {
+			return nil, pathError
+		}
+		var nfe *types.NotFound
+		if errors.As(err, &nfe) {
+			return nil, pathError
+		}
```

Comment:
Take it or leave it: We have a little utility to make this terser if you want
https://github.com/flipt-io/flipt/blob/2416a0149f1275db5d201ef5d8b8316e01c95b3c/errors/errors.go#L8-L19

```suggestion
		// try to return fs compatible error if possible
		if flipterrors.AsMatch[*types.NoSuchBucket](err) ||
			flipterrors.AsMatch[*type.NoSuchKey](err) ||
			flipterrors.AsMatch[*types.NotFound](err) {
			return nil, pathError
		}
```